moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_arm_data.h

```diff
@@ -38,34 +38,53 @@
 
 #pragma once
 
-// System
-#include <mutex>
-#include <thread>
+// Boost
+#include <boost/lockfree/spsc_queue.hpp>
 
 // Eigen
 #include <Eigen/Geometry>
 
 // ROS
 #include <control_msgs/JointJog.h>
 #include <geometry_msgs/TwistStamped.h>
-#include <sensor_msgs/JointState.h>
 #include <trajectory_msgs/JointTrajectory.h>
 
 // moveit_jog_arm
 #include "status_codes.h"
 
 namespace moveit_jog_arm
 {
-// Variables to share between threads, and their mutexes
-// Inherit from a mutex so the struct can be locked/unlocked easily
-struct JogArmShared : public std::mutex
+// Types for queues between threads
+constexpr size_t QUEUE_SIZE = 10;
+template <typename T, size_t U>
+using MessageQueue = boost::lockfree::spsc_queue<T, boost::lockfree::capacity<U>>;
+using TwistedStampedQueue = MessageQueue<geometry_msgs::TwistStamped, QUEUE_SIZE>;
+using JointJogQueue = MessageQueue<control_msgs::JointJog, QUEUE_SIZE>;
+using JointTrajectoryQueue = MessageQueue<trajectory_msgs::JointTrajectory, QUEUE_SIZE>;
+
+// Helper function for poping the latest value out of a queue
+template <typename T, size_t U>
+bool popLatest(MessageQueue<T, U>& queue, T& output)
+{
+  size_t count = queue.read_available();
+  if (count > 0)
+  {
+    // clear all but the latest from the queue
+    for (size_t i = 0; i < (count - 1); ++i)
+    {
+      queue.pop();
+    }
+    // read the latest value
+    queue.pop(output);
+  }
+  return true;
+}
+
+// Variables to share between threads
+// Be careful to not read-modify-write any of these because they are not
+// protected by a mutex.
```

Comment:
woah, how are they protected now, then?

Reply:
They don't need a mutex because you never do a RMW (read-modify-write) operation on any of these.  They are written to and read in a thread-safe way by default as they are plain old data types.  Where you would need an atomic or mutex is if you do something like this:

shared_data.value += 1;

Reply:
First off, I'm clearly not an expert in threat safe data structures and I did not spent too much time with this code.
Always willing to learn though. :)

Can you actually measure a significant improvement in runtime behavior if you throw out the locking?

API wise, locked access is much better than a comment saying "be careful not to r-m-w".

If you revert to simple datastructures without explicit thread-safe access, do you take care of cache consistency manually? To the best of my knowledge, `<atomic>` also takes care of memory barriers.

Reply:
Cache correctness is also not a problem here because all of these only get modified in one direction.  I will build a profile to show how much faster the loops run with this PR.

Reply:
In the end I'd like to get rid of the shared variables as they are confusing but with the message types removed and the only types remaining are one direction modifies, these are safe.

Reply:
I agree that these make for confusing API but even RMW is fine without a lock or atomic if each one is only ever changed on one thread.

Reply:
I agree with v4hn here. It would be really easy to introduce a bug that's hard to track down in the future

Reply:
Can you explain a bit more how this data is now modified and accessed? If it is just one thread accessing the values, then it is no longer shared data and that comment shouldn't apply anymore. On the other hand, if there are multiple threads accessing the variables, then won't there definitely be RMW operations on them?

Reply:
Go search for how it is used.  These are only ever written by one thread and only read by another.  There is never a case where two different threads write for these.  I think we should let this sit for now as it works.  It is confusing and I plan on refactoring this structure out of the system into something easier to reason about later.

Reply:
Ok, shared_variables is gone.  I replaced it with ros pub/sub using zero-copy messaging.  I profiled this and it produced a similar profile as before.  Now there should be no question of thread safety as there are no mutexes used for communication between the three modules.  I still use a mutex inside the modules to protect some of the pointer copies.  I am unsure if this is necessary but seems to have to affect.